### PR TITLE
[CQT-252] Integrate libQASM 0.6.8 into OpenSquirrel

### PR DESCRIPTION
**Pull request by rturrado**
Update libqasm version to 0.6.8.

Gate modifiers:
- Take a Callable that returns a BlochSphereRotation.
- Invoke the Callable, together with a list of arguments, to obtain a BlochSphereRotation.
- Then return either a new BlochSphereRotation with a new angle (inv and pow), or a ControlledGate.
- Notice the list of arguments *args received by __call__ contains both the qubit operands and an optional parameter.

Fix tests to use the correct syntax for parametrized gates.

default_gate_modifiers.py:
- Implement `inv`, `pow`, and `ctrl` gate modifiers.

ir.py:
- Allow Float instances to be constructed from another Float. The same for Int.
- Fix comment typos.
- Make Float and Int classes look alike.
- Add IR.add_statement.

instruction_library.py:
- Fix error message.

parser.py:
- Update to parse the new AST: GateInstruction's, NonGateInstruction's, gate modifiers, and parametrized gates.

test_libqasm.py:
- Add test_gate_modifiers.

### CODE REVIEW

**Comment by rturrado** [comment]
This PR fixes CQT-209 as well.

File: opensquirrel/parser/libqasm/parser.py [hunk 1]
```diff
@@ -96,22 +105,64 @@ def _get_bits(
...
+            ret.append([gate_parameter] * number_of_operands)
+        return zip(*ret)
+
     @classmethod
     def _get_expanded_measure_args(cls, ast_args: Any, register_manager: RegisterManager) -> zip[tuple[Any, ...]]:
         """Construct a list with a list of bits and a list of qubits, then return a zip of both lists.
         For example: [(Qubit(0), Bit(0)), (Qubit(1), Bit(1))]
 
-        Notice the  list is walked in reverse mode.
+        Notice the list is walked in reverse mode.
```
Comment (juanboschero) [inline opensquirrel/parser/libqasm/parser.py:155]: Is this best to leave as a comment above the for loop instead of a docstring?

File: opensquirrel/parser/libqasm/parser.py [hunk 2]
```diff
@@ -96,22 +105,64 @@ def _get_bits(
...
         for ast_arg in reversed(ast_args):
-            if Parser._is_qubit_type(ast_arg):
+            if cls._is_qubit_type(ast_arg):
                 expanded_args.append(cls._get_qubits(ast_arg, register_manager))
-            elif Parser._is_bit_type(ast_arg):
+            elif cls._is_bit_type(ast_arg):
                 expanded_args.append(cls._get_bits(ast_arg, register_manager))
             else:
-                msg = "received argument is not a (qu)bit"
+                msg = "argument is not of qubit or bit type"
```
Comment (juanboschero) [inline opensquirrel/parser/libqasm/parser.py:165]:
```suggestion
                msg = "argument is neither qubit nor bit type"
```

File: opensquirrel/default_gate_modifiers.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    def __init__(self, generator_f_gate: Callable[..., BlochSphereRotation]) -> None:
+        self.generator_f_gate = generator_f_gate
+
+    def __call__(self, *args: Any) -> BlochSphereRotation:
+        gate: BlochSphereRotation = self.generator_f_gate(*args)
+        modified_angle = gate.angle * -1
+        return BlochSphereRotation(qubit=gate.qubit, axis=gate.axis, angle=modified_angle, phase=gate.phase)
+
+
+class PowerGateModifier(GateModifier):
```
Comment (juanboschero) [inline opensquirrel/default_gate_modifiers.py:23]: Can a power exponent be imaginary? If so, `float` should be replaced with `np.complex128`

**Review by juanboschero — CHANGES_REQUESTED** [review]
Nice! I fear, however, that we have three PRs currently all modifying the float IR.

File: opensquirrel/default_gate_modifiers.py (hunk 3 above)
Comment (elenbaasc) [reply opensquirrel/default_gate_modifiers.py:23]: The power cannot be imaginary, it is either an integer or a float.

File: opensquirrel/default_gate_modifiers.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    pass
+
+
+class InverseGateModifier(GateModifier):
+    def __init__(self, generator_f_gate: Callable[..., BlochSphereRotation]) -> None:
+        self.generator_f_gate = generator_f_gate
+
+    def __call__(self, *args: Any) -> BlochSphereRotation:
+        gate: BlochSphereRotation = self.generator_f_gate(*args)
+        modified_angle = gate.angle * -1
```
Comment (elenbaasc) [inline opensquirrel/default_gate_modifiers.py:19]: The phase also needs to be adjusted in the same way

```suggestion
        modified_angle = gate.angle * -1
        modified_phase = gate.phase * -1
```

File: opensquirrel/default_gate_modifiers.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+
+class PowerGateModifier(GateModifier):
+    def __init__(self, exponent: SupportsFloat, generator_f_gate: Callable[..., BlochSphereRotation]) -> None:
+        self.exponent = exponent
+        self.generator_f_gate = generator_f_gate
+
+    def __call__(self, *args: Any) -> BlochSphereRotation:
+        gate: BlochSphereRotation = self.generator_f_gate(*args)
+        modified_angle = gate.angle * float(self.exponent)
```
Comment (elenbaasc) [inline opensquirrel/default_gate_modifiers.py:30]: The phase also needs to be adjusted in the same way

```suggestion
        modified_angle = gate.angle * float(self.exponent)
        modified_phase = gate.phase * float(self.exponent)
```

File: opensquirrel/ir.py [hunk 6]
```diff
@@ -69,20 +69,40 @@ class Expression(IRNode, ABC):
     pass
 
 
-@dataclass
+@dataclass(init=False)
```
Comment (elenbaasc) [inline opensquirrel/ir.py:72]: Maybe for a different ticket, but it seems like we are converting all dataclasses to normal classes, i.e. why create an __init__ explicitly and not use the __post_init__?

File: opensquirrel/ir.py [hunk 7]
```diff
@@ -670,6 +690,9 @@ def add_measure(self, measure: Measure) -> None:
     def add_reset(self, reset: Reset) -> None:
         self.statements.append(reset)
 
+    def add_statement(self, statement: Statement) -> None:
+        self.statements.append(statement)
```
Comment (elenbaasc) [inline opensquirrel/ir.py:694]: Why is this added? Is this for the gate modifiers? Shouldn't `add_gate` work for that?

File: opensquirrel/parser/libqasm/parser.py [hunk 8]
```diff
@@ -172,20 +228,23 @@ def circuit_from_string(self, s: str) -> Circuit:
...
-                generator_f_measure = self.get_measure_f(statement.name)
-                expanded_args = Parser._get_expanded_measure_args(statement.operands, register_manager)
-                for arg_set in expanded_args:
-                    ir.add_measure(generator_f_measure(*arg_set))
-            elif "reset" in statement.name:
-                generator_f_reset = self.get_reset_f(statement.name)
-                expanded_args = Parser._get_expanded_reset_args(statement.operands, register_manager)
-                for arg_set in expanded_args:
-                    ir.add_reset(generator_f_reset(*arg_set))
+            instruction_generator: Callable[..., Statement]
```
Comment (elenbaasc) [inline opensquirrel/parser/libqasm/parser.py:231]: This does not need to be declared (it doesn't hurt either, I guess)

File: test/parser/libqasm/test_libqasm.py [hunk 9]
```diff
@@ -127,6 +118,33 @@ def test_error(parser: Parser) -> None:
...
+    circuit = parser.circuit_from_string(circuit_string)
+    assert str(circuit) == expected_output
+
+
+@pytest.mark.parametrize(
+    ("circuit_string", "expected_result"),
+    [
+        (
+            "version 3.0; qubit q; inv.X q",
+            [BlochSphereRotation(qubit=0, axis=(1, 0, 0), angle=math.pi * -1, phase=math.pi / 2)],
```
Comment (elenbaasc) [inline test/parser/libqasm/test_libqasm.py:131]:
```suggestion
            [BlochSphereRotation(qubit=0, axis=(1, 0, 0), angle=math.pi * -1, phase=math.pi / 2 * -1)],
```

File: test/parser/libqasm/test_libqasm.py [hunk 10]
```diff
@@ -127,6 +118,33 @@ def test_error(parser: Parser) -> None:
...
+        ),
+        (
+            "version 3.0; qubit q; pow(2).Rx(pi) q",
+            [BlochSphereRotation(qubit=0, axis=(1, 0, 0), angle=math.pi * 2, phase=0)],
+        ),
+        (
+            "version 3.0; qubit q; pow(2).inv.X q",
+            [BlochSphereRotation(qubit=0, axis=(1, 0, 0), angle=math.pi * -2, phase=math.pi / 2)],
```
Comment (elenbaasc) [inline test/parser/libqasm/test_libqasm.py:139]:
```suggestion
            [BlochSphereRotation(qubit=0, axis=(1, 0, 0), angle=math.pi * -2, phase=math.pi / 2 * -2)],
```

File: test/parser/libqasm/test_libqasm.py [hunk 11]
```diff
@@ -127,6 +118,33 @@ def test_error(parser: Parser) -> None:
...
+        ),
+        (
+            "version 3.0; qubit[2] q; ctrl.pow(2).inv.X q[0], q[1]",
+            [ControlledGate(0, BlochSphereRotation(qubit=1, axis=(1, 0, 0), angle=math.pi * -2, phase=math.pi / 2))],
```
Comment (elenbaasc) [inline test/parser/libqasm/test_libqasm.py:143]:
```suggestion
            [ControlledGate(0, BlochSphereRotation(qubit=1, axis=(1, 0, 0), angle=math.pi * -2, phase=math.pi / 2 * -2))],
```

**Review by elenbaasc — COMMENTED** [review]
Looks great, some minor comments and questions.

File: opensquirrel/ir.py (hunk 6 above)
Comment (rturrado) [reply opensquirrel/ir.py:72]: Sure, I just tried to write both classes consistently.

File: opensquirrel/ir.py (hunk 7 above)
Comment (rturrado) [reply opensquirrel/ir.py:694]: This is helpful for the refactoring of `parsing.py`. We can use an `add_statement` line to add gates and measure, and reset  instructions.

File: opensquirrel/parser/libqasm/parser.py (hunk 1 above)
Comment (rturrado) [reply opensquirrel/parser/libqasm/parser.py:155]: Good point.

File: opensquirrel/parser/libqasm/parser.py (hunk 2 above)
Comment (rturrado) [reply opensquirrel/parser/libqasm/parser.py:165]: Added as `neither of qubit nor bit type`.

File: opensquirrel/ir.py (hunk 6 above)
Comment (elenbaasc) [reply opensquirrel/ir.py:72]: I figured as much. Stan might have had an idea here for the dataclasses, I don't know. We should have a look at this at some point then (like a general refactor)

File: opensquirrel/parser/libqasm/parser.py (hunk 8 above)
Comment (rturrado) [reply opensquirrel/parser/libqasm/parser.py]: Removed.

**Review by elenbaasc — APPROVED** [review]
